lib/private/Share20/Manager.php

```
@@ -675,6 +677,104 @@ public function createShare(\OCP\Share\IShare $share) {
+		}
+
+		/**
+		 * If the share was already shared with new owner, then we can delete it
+		 */
+		if ($share->getSharedWith() === $newOwner) {
+			// Unmount the shares before deleting, so we don't try to get the storage later on.
+			$shareMountPoint = $this->mountManager->find('/' . $newOwner . '/files' . $share->getTarget());
+			if ($shareMountPoint) {
+				$this->mountManager->removeMount($shareMountPoint->getMountPoint());
+			}
+			$this->deleteShare($share);
+		} else {
+			$sharedWith = $share->getSharedWith();
+			if (!isset($extraArgs['finalTarget'])) {
+				throw new \Exception('finalTarget is not set');
+			}
+
+			$targetFile = '/' . rtrim(basename($extraArgs['finalTarget']), '/') . '/' . ltrim(basename($share->getTarget()), '/');
+			/**
+			 * Scenario where share is made by old owner to a user different
+			 * from new owner
+			 */
+			if (($sharedWith !== null) && ($sharedWith !== $oldOwner) && ($sharedWith !== $newOwner)) {
+				$sharedBy = $share->getSharedBy();
+				$sharedOwner = $share->getShareOwner();
+				//The origin of the share now has to be the destination user.
+				if (($sharedBy === $oldOwner) && ($sharedOwner === $oldOwner)) {
+					$share->setSharedBy($newOwner);
+					$share->setShareOwner($newOwner);
+					$share->setTarget($targetFile);
+				}
+			} else {
+				$oldShare = $share;
+				if ($share->getShareOwner() === $oldOwner) {
+					$share->setShareOwner($newOwner);
+				}
+				if ($share->getSharedBy() === $oldOwner) {
+					$share->setSharedBy($newOwner);
+				}
+			}
+			/*
+			 * If the share is already moved then updateShare would cause exception
+			 * This can happen if the folder is shared and file(s) inside the folder
+			 * has shares, for example public link
+			 */
+			if ($share->getShareType() === \OCP\Share::SHARE_TYPE_LINK) {
+				$sharePath = ltrim($share->getNode()->getPath(), '/');
+				if (strpos($sharePath, $extraArgs['finalTarget']) !== false) {
```

This needs clarification: what if the `$sharePath` is `a/b/c` and the final target is `c`? I guess it should throw an exception in this case